browser/request_otr/request_otr_browsertest.cc

```diff
@@ -0,0 +1,425 @@
+  auto* contents = model->GetActiveWebContents();
+  auto* infobar_manager =
+      infobars::ContentInfoBarManager::FromWebContents(contents);
+  TestObserver observer;
+  // Set up expectation that an infobar will NOT appear later.
+  EXPECT_CALL(observer, OnInfoBarAdded(_)).Times(0);
+  infobar_manager->AddObserver(&observer);
+
+  SetRequestOTRPref(RequestOTRService::RequestOTRActionOption::kAsk);
+  GURL url = embedded_test_server()->GetURL("sensitive.a.com", "/simple.html");
+  NavigateTo(url);
+  ASSERT_TRUE(IsShowingInterstitial());
+
+  // Simulate click on 'Proceed Normally' button. This should navigate to
+  // the originally requested page in off-the-record mode.
+  ClickAndWaitForNavigation("back-button");
+  ASSERT_FALSE(IsShowingInterstitial());
+  // Request-OTR infobar should never appear, because the user requested to
+  // proceed normally, so we should not be in off-the-record mode.
+
+  infobar_manager->RemoveObserver(&observer);
+}
+
+IN_PROC_BROWSER_TEST_F(RequestOTRBrowserTest, IncludeExclude) {
+  ASSERT_TRUE(InstallMockExtension());
+  SetRequestOTRPref(RequestOTRService::RequestOTRActionOption::kAsk);
+  GURL url1 = embedded_test_server()->GetURL("www.b.com", "/simple.html");
+  NavigateTo(url1);
+  ASSERT_TRUE(IsShowingInterstitial());
+  GURL url2 =
+      embedded_test_server()->GetURL("notsensitive.b.com", "/simple.html");
+  NavigateTo(url2);
+  ASSERT_FALSE(IsShowingInterstitial());
+}
+
+IN_PROC_BROWSER_TEST_F(RequestOTRBrowserTest, HistoryAfterStandardNavigation) {
+  ASSERT_EQ(GetHistoryCount(), 0);
+  NavigateTo(
+      embedded_test_server()->GetURL("notsensitive.b.com", "/simple.html"));
+  ASSERT_EQ(GetHistoryCount(), 1);
+}
+
+IN_PROC_BROWSER_TEST_F(RequestOTRBrowserTest, HistoryAfterOTRNavigation) {
+  ASSERT_TRUE(InstallMockExtension());
+
+  ASSERT_EQ(GetHistoryCount(), 0);
+  SetRequestOTRPref(RequestOTRService::RequestOTRActionOption::kAlways);
+  NavigateTo(embedded_test_server()->GetURL("sensitive.a.com", "/simple.html"));
+  ASSERT_EQ(GetHistoryCount(), 0);
+}
+
+IN_PROC_BROWSER_TEST_F(RequestOTRBrowserTest,
+                       WindowOpenAfterStandardNavigation) {
+  NavigateTo(embedded_test_server()->GetURL("sensitive.a.com", "/simple.html"));
+  ASSERT_TRUE(content::ExecJs(
+      web_contents(), "window.open('notsensitive.b.com/simple.html');"));
+  ASSERT_NE(content::EvalJs(web_contents(), "window.opener"), nullptr);
+}
+
+IN_PROC_BROWSER_TEST_F(RequestOTRBrowserTest, WindowOpenAfterOTRNavigation) {
+  ASSERT_TRUE(InstallMockExtension());
+
+  // Always use request-otr for sensitive sites (skipping interstitial).
+  SetRequestOTRPref(RequestOTRService::RequestOTRActionOption::kAlways);
+
+  NavigateTo(embedded_test_server()->GetURL("sensitive.a.com", "/simple.html"));
+  ASSERT_TRUE(content::ExecJs(
+      web_contents(), "window.open('notsensitive.b.com/simple.html');"));
+  ASSERT_EQ(content::EvalJs(web_contents(), "window.opener"), nullptr);
```

Comment:
I'm not clear on what we're testing here because it will always be null for a different domain. We should be testing that a normally scriptable window (same domain) has an opener without OTR and does not have one with OTR

Reply:
added same-domain tests